Pass in parallel strategy to tp_transform API

feikou opened the pull request:
Summary: Support passing in parallel strategy map and apply TP transform based on that. This is to make it easier manually select layers from real model to parallelize and benchmark.

Test Plan: buck test mode/opt  -c fbcode.enable_gpu_sections=true //caffe2/test/distributed/_tensor/experimental:tp_transform

Differential Revision: D50591039

wanchaol commented on torch/distributed/_tensor/experimental/tp_transform.py:
```diff
@@ -65,28 +72,58 @@ def __init__(
+    parameter_placements: Dict[str, Placement] = {}
+    for linear_fqn, parallel_style in parallel_strategies.items():
+        weight_fqn = f"{linear_fqn}.weight"
+        bias_fqn = f"{linear_fqn}.bias"
```
I think because linear does not always have a `bias` parameter, we should check if the linear_fqn passed in have bias or not?

Or we can return the placement here but later when we apply the placements to the linear related parameters, we check if there's a bias parameter at that time.

wanchaol commented on test/distributed/_tensor/experimental/test_tp_transform.py:
```diff
@@ -42,6 +49,13 @@ def test_tensor_parallel_transformation_e2e(self):
         torch.manual_seed(0)
         model = MLPListModule(2).to(device=self.device_type)
         inputs = (torch.randn((10, 12)).to(device=self.device_type),)
+        parallel_strategies: Dict[str, ParallelStyle] = {
```
can you also add a test where the mlps are created with `bias=False`? To verify that with no bias the pass should also work

feikou replied:
Absolutely! Good catch

wanchaol reviewed (commented):
Looks good, we should make sure when `bias=False` the transformation also works

XilunWu commented on test/distributed/_tensor/experimental/test_tp_transform.py:
```diff
@@ -37,11 +44,57 @@ class TensorParallelTest(DTensorTestBase):
+        self.assertEqual(res, tp_res)
+
+    @with_comms
+    def test_tp_transform_no_bias(self):
```
LGTM! Thx for addressing Wanchao's comment!

wanchaol commented on test/distributed/_tensor/experimental/test_tp_transform.py:
```diff
@@ -37,11 +44,57 @@ class TensorParallelTest(DTensorTestBase):
         model = MLPListModule(2).to(device=self.device_type)
         inputs = (torch.randn((10, 12)).to(device=self.device_type),)
+        parallel_strategies: Dict[str, ParallelStyle] = {
+            "mlps.0.0": ColwiseParallel,
```
usually in eager mode we do `ColwiseParallel()`, but I guess for graph transform either way is fine as it's just a tag

wanchaol reviewed (approved):
lgtm!

feikou committed:
Pass in parallel strategy to tp_transform API (#112286)

Summary:

Support passing in parallel strategy map and apply TP transform based on that. This is to make it easier manually select layers from real model to parallelize and benchmark.

Test Plan: buck test mode/opt  -c fbcode.enable_gpu_sections=true //caffe2/test/distributed/_tensor/experimental:tp_transform

Reviewed By: wanchaol

Differential Revision: D50591039
